Readme.md
![Build](https://github.com/palkan/influx_udp/workflows/Test/badge.svg)
[![Hex Version](https://img.shields.io/hexpm/v/influx_udp.svg)](https://hex.pm/packages/influx_udp)

# Erlang InfluxDB UDP Writer

Write data to [InfluxDB](http://influxdb.com) (>= 0.9) via UDP (see [InfluxDB docs](https://docs.influxdata.com/influxdb/v1.7/supported_protocols/udp/)).

**Erlang/OTP version:** >=17.1

## Setup

rebar.config
```erlang
%% using Hex
{deps, [
  influx_udp
]}.

%% from source
{deps, [
  {influx_udp, ".*", {git, "https://github.com/palkan/influx_udp.git", "master"}}
]}.
```

app.config
```erlang
[
  {influx_udp,
    [
      {influx_host, '127.0.0.1'},
      {influx_port, 8089},
      {pool_size, 5}, %% defaults to 3
      {max_overflow, 10} %% defaults to 1
    ]
  }
].
```

## Usage

First, you need to start the application:

```erlang
influx_udp:start().
```

Now you can create _pools_ and write data to InfluxDB.

## Pools

The default pool is started on application start if you specified `influx_host` and `influx_port` in the configuration file (see above).

You can run pools manually:

```erlang
influx_udp:start_pool(my_pool, #{ host => 'yet.another.influx.host' }).
```

Options not specified in `influx_udp:start_pool/2` would be taken from the default configuration.

## Writing data

```erlang

%% Writing to the named pool with tags
influx_udp:write_to(
  my_pool,
  Series::string()|atom()|binary(), Points::list(map())|list(proplists:proplist())|map()|proplists:proplist(),
  Tags::proplists:proplist()|map()).

influx_udp:write_to(my_pool, "cpu", [{value, 88}], [{host, 'eu-west'}]).

%% Writing to default pool
influx_udp:write("cpu", [#{value => 88}, #{value => 22}, #{value => 33}], [{host, 'eu-west'}]).

%% Writing data with time
influx_udp:write("cpu", #{value => 88}, #{host => 'eu-west'}, 1434055562000000000).

%% or with current time
influx_udp:write("cpu", #{value => 88}, #{host => 'eu-west'}, true).

%% Writing to default pool without tags
influx_udp:write(Series, Points).

%% Writing raw valid InfluxDB input data
influx_udp:write(Data::binary()).

%% or
influx_udp:write_to(my_pool, Data::binary()).

%% Write Influx-valid map or proplist
influx_udp:write(#{ measurement => test, fields => #{ val => 1} }).

%% or many points
influx_udp:write(
  #{ measurement => test, fields => #{ val => 1} },
  #{ measurement => test2, fields => #{ val => 2}, tags => { host => test}}
)
```

## Encoder

Module `influx_line` provides methods to encode erlang terms to Line protocol.
Encoder automatically sets timestamps (unique) when encoding list of points (see below).

```erlang

%% convert map or proplist to line
influx_line:encode(#{ measurement => test, fields => #{ val => 1} }).

#=> <<"test val=1">>

%% convert list of points to lines
influx_line:encode([
  #{ measurement => test, fields => #{ val => 1} },
  #{ measurement => test2, fields => #{ val => 2}, tags => { host => test}}
]).

#=> <<"test val=1 1434305562895000000\ntest2,host=test val=2 1434305562895000001">>

%% convert any map/proplist to line
influx_line:encode(test, #{ val => 1}).

#=> <<"test val=1">>

%% convert many points with the same measurement and tags to line
influx_line:encode(test, [#{ val => 1}, #{ val => 2}], #{ host => test}, 100).

#=> <<"test,host=test val=1 100\ntest,host=test val=2 101\n">>
```

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/palkan/influx_udp.

## License

The library is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


Changelog.md
# Change log

## master

## 1.1.2 (2021-10-30) 🎃

- Fix type specs.

## 1.1.1 (2020-12-29)

- Fix Erlang Logger compatibility.

## 1.1.0 (2020-08-17)

- Drop Lager dependency.

Now standard Logger or `error_logger` (for OTP <21) is used instead.

[PR](https://github.com/palkan/influx_udp/pull/16).

## 1.0.0 (2019-01-06)

[Hex package](https://hex.pm/packages/influx_udp) is available 🎉.

- Default pool behaviour changed.

Previously, the default configuration had both `influx_host` and `influx_port` defined, which made
the application start the _default_ pool connecting to `127.0.0.1:4444` (which didn't make a lot of sense, did it?).

We also had a special `{default_pool, false}` configuration variable to disable the default pool.

We removed the default value for `influx_host` (and changed the default `influx_port` to **8089**).

Now if `influx_host` is not defined, we do not start the default pool.

## 0.9.1 (2015-06-14)

- Add Line encoder

## 0.9.0 (2015-06-14)

- Add support for InfluxDB 0.9 (tags, line protocol)

## 0.8.0 (2015-06-13)

- Add pools functionality

Major and minor versions now follows InfluxDB version.

## 0.1.0 (2014-12-17)

- Basic functionality.


test/influx_udp_test.erl
-module(influx_udp_test).
-include_lib("influx_udp/include/influx_udp_priv.hrl").
-include_lib("influx_udp/include/influx_udp.hrl").
-include_lib("eunit/include/eunit.hrl").

-define(setup(F), {setup, fun setup_/0, fun cleanup_/1, F}).

%% The default influx_udp port
-define(PORT, 8089).
-define(PORT2, 44515).

setup_() ->
  {ok, UDP1} = test_udp_server:start(?PORT),
  {ok, UDP2} = test_udp_server:start(?PORT2),

  %% Set default configuration
  ulitos_app:set_var(?APP, influx_host, '127.0.0.1'),

  influx_udp:start(),
  {ok, _Pid} = influx_udp:start_pool(test, #{ pool_size => 1, port => ?PORT2}),
  {UDP1, UDP2}.

cleanup_({UDP1, UDP2}) ->
  influx_udp:stop(),
  test_udp_server:stop(UDP1),
  test_udp_server:stop(UDP2).

write_point_test_() ->
  [{"Write point",
    ?setup(
      fun(Config) ->
        {inorder,
          [
            write_point_t_(Config)
          ]
        }
      end
    )
  }].

write_point_pool_test_() ->
  [{"Write point to named pool",
    ?setup(
      fun(Config) ->
        {inorder,
          [
            write_point_to_pool_t_(Config)
          ]
        }
      end
    )
  }].

write_influx_point_test_() ->
  [{"Write point to named pool",
    ?setup(
      fun(Config) ->
        {inorder,
          [
            write_influx_point_t_(Config)
          ]
        }
      end
    )
  }].


write_point_t_({UDP1, _UDP2}) ->
  influx_udp:write(test, [{val, 1}], [{host, test}], 100),
  timer:sleep(1),
  [
    ?_assertEqual(<<"test,host=test val=1 100">>, gen_server:call(UDP1, msg))
  ].

write_point_to_pool_t_({_UDP1, UDP2}) ->
  influx_udp:write_to(test, <<"test_pool">>),
  timer:sleep(1),
  [
    ?_assertEqual(<<"test_pool">>, gen_server:call(UDP2, msg))
  ].

write_influx_point_t_({UDP1, _UDP2}) ->
  influx_udp:write([
    #{ measurement => <<"test">>, fields => #{ val => 1}, time => 1},
    #{ measurement => <<"test2">>, fields => #{ val => 2}, time => 2}
  ]),
  timer:sleep(1),
  [
    ?_assertEqual(<<"test val=1 1\ntest2 val=2 2\n">>, gen_server:call(UDP1, msg))
  ].


test/test_udp_server.erl
-module(test_udp_server).
-behaviour(gen_server).
-include_lib("influx_udp/include/influx_udp_priv.hrl").
-include_lib("influx_udp/include/influx_udp.hrl").
-define(SERVER, ?MODULE).

%% ------------------------------------------------------------------
%% API Function Exports
%% ------------------------------------------------------------------

-export([start/1, stop/1]).

%% ------------------------------------------------------------------
%% gen_server Function Exports
%% ------------------------------------------------------------------

-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-record(state,{
  sock,
  msg
}).

%% ------------------------------------------------------------------
%% API Function Definitions
%% ------------------------------------------------------------------

start(Port) ->
  gen_server:start(?MODULE, [Port], []).

stop(Pid) ->
  Pid ! stop.

%% ------------------------------------------------------------------
%% gen_server Function Definitions
%% ------------------------------------------------------------------

init([Port]) ->
  ?D({start_test_server, Port}),
  {ok, Socket} = gen_udp:open(Port, [binary, {active, once}]),
  ?D({test_server_start_listening}),
  {ok, #state{sock=Socket}}.

handle_call(msg, _, #state{msg=Msg}=State) ->
  {reply, Msg, State};

handle_call(_Request, _From, State) ->
  {reply, ok, State}.

handle_cast(_Msg, State) ->
  {noreply, State}.

handle_info({udp, Socket, _Host, _Port, Bin}, #state{sock=Socket}=State) ->
  ?D({message_received, Bin}),
  inet:setopts(Socket, [{active, once}]),
  {noreply, State#state{msg=Bin}};

handle_info(stop, State) ->
  {stop, normal, State};

handle_info(_Info, State) ->
  {noreply, State}.

terminate(_Reason, #state{sock=Sock}) ->
  gen_udp:close(Sock),
  ok.

code_change(_OldVsn, State, _Extra) ->
  {ok, State}.

%% ------------------------------------------------------------------
%% Internal Function Definitions
%% ------------------------------------------------------------------


test/src/test_udp_app.erl
-module(test_udp_app).

-behaviour(application).

%% Application callbacks
-export([start/2, stop/1]).

%% ===================================================================
%% Application callbacks
%% ===================================================================

start(_StartType, _StartArgs) ->
  test_udp_sup:start_link().

stop(_State) ->
  ok.


src/influx_udp_sup.erl
-module(influx_udp_sup).
-behaviour(supervisor).
-include_lib("influx_udp/include/influx_udp_priv.hrl").

%% API
-export([start_link/0, start_pool/2]).

%% Supervisor callbacks
-export([init/1]).

%% Helper macro for declaring children of supervisor
-define(
  CHILD(I, Type), {I, {I, start_link, []}, permanent, 5000, Type, [I]}
).

%% ===================================================================
%% API functions
%% ===================================================================

start_link() ->
  supervisor:start_link({local, ?MODULE}, ?MODULE, []).

-spec start_pool(
    Name::atom(),
    Options::map()
  ) -> {ok, undefined | pid()} |
       {ok, undefined | pid(), any()} |
       {error, any()}.
start_pool(Name, Options) ->
  SizeArgs = [
    {size, maps:get(pool_size, Options)},
    {max_overflow, maps:get(max_overflow, Options)}
  ],

  PoolArgs = [
    {name, {local, Name}},
    {worker_module, influx_udp_worker}
  ] ++ SizeArgs,

  supervisor:start_child(
    ?MODULE,
    poolboy:child_spec(Name, PoolArgs, [{options, Options}])
  ).

%% ===================================================================
%% Supervisor callbacks
%% ===================================================================

init([]) ->
  Children = [
    ?CHILD(influx_udp, worker)
  ],
  {ok, { {one_for_one, 5, 10}, Children} }.


src/influx_udp_worker.erl
-module(influx_udp_worker).
-behaviour(gen_server).
-behaviour(poolboy_worker).
-include_lib("influx_udp/include/influx_udp_priv.hrl").
-include_lib("influx_udp/include/influx_udp.hrl").
-define(SERVER, ?MODULE).

%% ------------------------------------------------------------------
%% API Function Exports
%% ------------------------------------------------------------------

-export([start_link/1]).

%% ------------------------------------------------------------------
%% gen_server Function Exports
%% ------------------------------------------------------------------

-export([init/1, handle_call/3, handle_cast/2, handle_info/2,
         terminate/2, code_change/3]).

-record(state, {
  socket ::gen_udp:socket()|undefined,
  port ::inet:port_number()|undefined,
  host ::inet:ip_address()|inet:ip_hostname()|undefined,
  local_port ::inet:port_number()|undefined,
  debug = false ::boolean()
}).

%% ------------------------------------------------------------------
%% API Function Definitions
%% ------------------------------------------------------------------

start_link(Args) ->
  gen_server:start_link(?MODULE, Args, []).

%% ------------------------------------------------------------------
%% gen_server Function Definitions
%% ------------------------------------------------------------------
init([{options, #{ host := Host, port := Port} = _Options}]) ->
  AddrFamily = addr_family(Host),
  {ok, Socket} = gen_udp:open(0, [binary, {active, false}, AddrFamily]),
  {ok, LocalPort} = inet:port(Socket),
  ?I(#{msg => "Open UDP socket on port", port => LocalPort}),
  {
    ok,
    #state{
      socket = Socket,
      host = Host,
      port = Port,
      local_port = LocalPort,
      debug = ?Config(debug, false)
    }
  }.

handle_call({write, Bin}, _From, State) when is_binary(Bin) ->
  {reply, send_data(State, Bin), State};

handle_call({write, Points}, _From, State) ->
  {reply, send_data(State, influx_line:encode(Points)), State};

handle_call({write, Series, Points}, _From, State) ->
  {reply, send_data(State, influx_line:encode(Series, Points)), State};

handle_call({write, Series, Points, Tags}, _From, State) ->
  {reply, send_data(State, influx_line:encode(Series, Points, Tags)), State};

handle_call({write, Series, Points, Tags, Time}, _From, State) ->
  {reply, send_data(State, influx_line:encode(Series, Points, Tags, Time)), State};

handle_call(_Request, _From, State) ->
  {reply, ok, State}.

handle_cast(_Msg, State) ->
  {noreply, State}.

handle_info(_Info, State) ->
  {noreply, State}.

terminate(_Reason, #state{socket=Socket}) ->
  gen_udp:close(Socket),
  ok.

code_change(_OldVsn, State, _Extra) ->
  {ok, State}.

%% ------------------------------------------------------------------
%% Internal Function Definitions
%% ------------------------------------------------------------------

addr_family({_, _, _, _}) -> inet;
addr_family({_, _, _, _, _, _, _, _}) -> inet6.

send_data(_, {error, Reason} = Error) ->
  ?E(#{error => Reason}),
   Error;

send_data(#state{socket=Socket, port=Port, host=Host, debug = Debug}, Data) ->
  debug_send(Data, Debug),
  gen_udp:send(Socket, Host, Port, Data).

debug_send(_, false) -> ok;

debug_send(Bin, _) ->
  ?D(#{event => send_binary_data, size => byte_size(Bin), data => Bin}).


src/influx_line.erl
%%% @doc
%%% InfluxDB Line protocol encoder.
%%% @end
-module(influx_line).
-include_lib("influx_udp/include/influx_udp_priv.hrl").
-include_lib("influx_udp/include/influx_udp.hrl").

-define(Q, <<"\"">>).
-define(SPACE, <<" ">>).
-define(NEW_LINE, <<"\n">>).
-define(EMPTY, <<"">>).
-define(COMMA, <<",">>).
-define(EQ, <<"=">>).

-export([
  encode/1,
  encode/2,
  encode/3,
  encode/4,
  encode_tags/1,
  encode_fields/1,
  timestamp/0
]).

%% ------------------------------------------------------------------
%% API Function Definitions
%% ------------------------------------------------------------------

%% @doc
%% Encode InfluxDB data point (or list of points) to line.
%% Point should contain 'measurement' and 'fields' keys (as atoms).
%% Other available keys: 'tags', 'time'.
%% Return binary line(-s) or `{error, invalid_data}' if any point contains
%% neither 'measurement' nor 'fields' key.
%%
%% Example:
%%
%% ```
%%   influx_line:encode(
%%     #{
%%       measurement => cpu,
%%       fields => #{ value => 43},
%%       tags => #{ host => 'eu-west', ip => "127.0.0.1" },
%%       time => 10000000000
%%     }
%%   ).
%%   <<"cpu,host=eu-west,ip="127.0.0.1" value=43 10000000000\n">>
%% '''
%% @end
-spec encode(Points::influx_data_points()) -> Line::binary()|{error, Reason::atom()}.
encode(#{} = Data) ->
  encode_with_time(Data, undefined);

encode([{_, _}|_] = Data) ->
  encode_with_time(Data, undefined);

encode(List) when is_list(List) ->
  ?D(encode_list),
  encode_points_list(undefined, List, timestamp(), []);

encode(_Other) ->
  ?D({unknown_data, _Other}),
  {error, invalid_data}.

%% @doc
%% Encode measurement and fields (or list of fields) to line.
%% @end
-spec encode(
  Name::string()|atom()|binary(),
  Fields::influx_data_points()
) -> Line::binary() | {error, Reason::atom()}.
encode(Name, Fields) ->
  encode(Name, Fields, [], undefined).

%% @doc
%% Encode measurement, fields (or list of fields) and tags to line.
%% @end
-spec encode(
  Name::string()|atom()|binary(),
  Fields::influx_data_points(),
  Tags::influx_data_point()|binary()
) -> Line::binary() | {error, Reason::atom()}.
encode(Name, Fields, Tags) ->
  encode(Name, Fields, Tags, undefined).

%% @doc
%% Encode measurement, fields (or list of fields), tags and time to line.
%% If Time is integer then it's used as point time.
%% If Time is `true' then `inlux_line:timestamp()' is used as point time.
%% If there are several points, then every point is set a uniq time
%% (by incrementing provided time).
%%
%% Note: if there are several points and no time specified then current time
%% is used as base time.
%%
%% Example:
%%
%% ```
%%   %% several points (uniq times)
%%   influx_line:encode(
%%     cpu,
%%     [
%%       #{ value => 43 },
%%       #{ value => 12 }
%%     ],
%%     #{ host => 'eu-west', ip => "127.0.0.1" },
%%     1000000000
%%   ).
%%   <<"cpu,host=eu-west,ip="127.0.0.1" value=43 1000000000\n
%%      cpu,host=eu-west,ip="127.0.0.1" value=12 1000000001\n">>
%%
%%   %% one point without time
%%   influx_line:encode(
%%     cpu,
%%     [
%%       #{ value => 43 },
%%       #{ value => 12 }
%%     ],
%%     #{ host => 'eu-west', ip => "127.0.0.1" }
%%   ).
%%   <<"cpu,host=eu-west,ip="127.0.0.1" value=43\n">>
%% '''
%% @end
-spec encode(
  Name::string()|atom()|binary(),
  Points::list(influx_data_point())|influx_data_point(),
  Tags::influx_data_point()|binary(),
  Time::non_neg_integer()|atom()
) -> Line::binary() | {error, Reason::atom()}.
encode(Name, #{} = Map, Tags, Time) ->
  encode_item(Name, Map, Tags, Time);

encode(Name, [{_, _}|_] = List, Tags, Time) ->
  encode_item(Name, List, Tags, Time);

encode(Name, [[{_, _}|_]|_] = Points, Tags, Time) ->
  encode_list(Name, Points, Tags, Time);

encode(Name, [#{}|_] = Points, Tags, Time) ->
  encode_list(Name, Points, Tags, Time);

encode(_, _, _, _) -> {error, invalid_data}.

%% @doc
%% Returns current UTC time in nanoseconds
%% @end
-spec timestamp() -> Time::non_neg_integer().
timestamp() ->
  ulitos:timestamp()*1000000.

%% ------------------------------------------------------------------
%% Internal Function Definitions
%% ------------------------------------------------------------------
encode_item(Name, Item, Tags, Time) ->
  BTags = encode_tags(Tags),
  Measurement = to_key(Name),
  Prefix = <<Measurement/binary, BTags/binary>>,
  encode_item_with_prefix(Prefix, Item, Time).

encode_item_with_prefix(Prefix, Item, Time) ->
  BTime = encode_time(Time),
  Fields = encode_fields(Item),
  concat_line(Prefix, Fields, BTime).

encode_list(Name, [Item], Tags, Time) ->
  encode_item(Name, Item, Tags, Time);

encode_list(Name, List, Tags, Time) when is_integer(Time) ->
  BTags = encode_tags(Tags),
  Measurement = to_key(Name),
  Prefix = <<Measurement/binary, BTags/binary>>,
  encode_list_with_time(undefined, List, Prefix, Time, []);

encode_list(Name, List, Tags, _Time) ->
  encode_list(Name, List, Tags, influx_line:timestamp()).

encode_list_with_time(Item, [], _, _, Acc) ->
  lists:foldl(fun concat_lines/2, ?EMPTY, [Item|Acc]);

encode_list_with_time(undefined, [H|Rest], Prefix, Time, Acc) ->
  encode_list_with_time(
    encode_item_with_prefix(Prefix, H, Time),
    Rest,
    Prefix,
    Time + 1,
    Acc
  );

encode_list_with_time(Item, [H|Rest], Prefix, Time, Acc) ->
  encode_list_with_time(
    encode_item_with_prefix(Prefix, H, Time),
    Rest,
    Prefix,
    Time + 1,
    [Item|Acc]
  ).

encode_points_list({error, _} = Error, _, _, _) ->
  Error;

encode_points_list(Item, [], _, Acc) ->
  lists:foldl(fun concat_lines/2, ?EMPTY, [Item|Acc]);

encode_points_list(undefined, [H|Rest], Time, Acc) ->
  encode_points_list(
    encode_with_time(H, Time),
    Rest,
    Time + 1,
    Acc
  );

encode_points_list(Item, [H|Rest], Time, Acc) ->
  encode_points_list(
    encode_with_time(H, Time),
    Rest,
    Time + 1,
    [Item|Acc]
  ).

encode_with_time([{_, _}|_] = List, BaseTime) ->
  M = proplists:get_value(measurement, List, undefined),
  F = proplists:get_value(fields, List, undefined),
  if M =:= undefined orelse F =:= undefined
    -> {error, invalid_data};
    true ->
      Time = encode_time(
        proplists:get_value(time, List, BaseTime)
      ),
      Tags = encode_tags(
        proplists:get_value(tags, List, [])
      ),
      Measurement = to_key(M),
      Fields = encode_fields(F),
      concat_line(Measurement, Tags, Fields, Time)
  end;

encode_with_time(#{measurement := M, fields := F} = Map, BaseTime) ->
  Time = encode_time(
    maps:get(time, Map, BaseTime)
  ),
  Tags = encode_tags(
    maps:get(tags, Map, [])
  ),
  Measurement = to_key(M),
  Fields = encode_fields(F),
  concat_line(Measurement, Tags, Fields, Time);

encode_with_time(_, _) -> {error, invalid_data}.

encode_time(undefined) ->
  ?EMPTY;

encode_time(true) ->
  encode_time(timestamp());

encode_time(N) ->
  to_time_val(N).

encode_tags(Tags) when is_binary(Tags) ->
  Tags;

encode_tags(#{} = Map) ->
  remove_trailing_comma(
    maps:fold(fun encode_map_tags/3, ?COMMA, Map)
  );

encode_tags(List) when is_list(List) ->
  remove_trailing_comma(
    lists:foldl(fun encode_list_tags/2, ?COMMA, List)
  );

encode_tags(_) -> ?SPACE.

encode_fields(#{} = Map) ->
  remove_trailing_comma(
    maps:fold(fun encode_map_fields/3, ?SPACE, Map)
  );

encode_fields(List) when is_list(List) ->
  remove_trailing_comma(
    lists:foldl(fun encode_list_fields/2, ?SPACE, List)
  );

encode_fields(_) -> {error, invalid_data}.

encode_map_fields(K, V, Acc) ->
  BK = to_key(K),
  BV = to_val(V),
  <<Acc/binary, BK/binary, ?EQ/binary, BV/binary, ?COMMA/binary>>.

encode_list_fields({K, V}, Acc) ->
  encode_map_fields(K, V, Acc).

encode_map_tags(K, V, Acc) ->
  BK = to_key(K),
  BV = to_tag_val(V),
  <<Acc/binary, BK/binary, ?EQ/binary, BV/binary, ?COMMA/binary>>.

encode_list_tags({K, V}, Acc) ->
  encode_map_tags(K, V, Acc).

to_key(A) when is_atom(A) ->
  to_key(atom_to_list(A));

to_key(S) when is_list(S) ->
  list_to_binary(S);

to_key(N) when is_integer(N) ->
  integer_to_binary(N);

to_key(S) -> S.

to_val(true) ->
  <<"true">>;

to_val(false) ->
  <<"false">>;

to_val(N) when is_integer(N) ->
  integer_to_binary(N);

to_val(N) when is_float(N) ->
  float_to_binary(N);

to_val(A) when is_atom(A) ->
  to_val(atom_to_list(A));

to_val(S) when is_list(S) ->
  to_val(list_to_binary(S));

to_val(S) when is_binary(S) ->
  Escaped = escape_string(S),
  <<?Q/binary, Escaped/binary, ?Q/binary>>.

to_tag_val(V) when is_integer(V) ->
  to_val(V);

to_tag_val(V) when is_float(V) ->
  to_val(V);

to_tag_val(V) when is_atom(V) ->
  to_tag_val(atom_to_list(V));

to_tag_val(V) when is_list(V) ->
  to_tag_val(list_to_binary(V));

to_tag_val(V) when is_binary(V) ->
  escape_string(V).

to_time_val(V) ->
  B = to_tag_val(V),
  <<?SPACE/binary, B/binary>>.

escape_string(Tag) ->
  binary:replace(Tag,
    [?COMMA, ?Q, ?SPACE, ?EQ],
    <<"\\">>,
    [global, {insert_replaced, 1}]
  ).

concat_line(A, B, C) ->
  <<A/binary, B/binary, C/binary>>.

concat_line(A, B, C, D) ->
  <<A/binary, B/binary, C/binary, D/binary>>.

concat_lines(Acc, B) ->
  concat_line(Acc, ?NEW_LINE, B).

remove_trailing_comma(?EMPTY) -> ?EMPTY;

remove_trailing_comma(B) -> binary:part(B, {0, byte_size(B) - 1}).

-ifdef(TEST).
-include_lib("eunit/include/eunit.hrl").

to_key_test() ->
  ?assertEqual(<<"test">>, to_key(test)),
  ?assertEqual(<<"test">>, to_key("test")),
  ?assertEqual(<<"test">>, to_key(<<"test">>)),
  ?assertEqual(<<"123">>, to_key(123)).

to_val_test() ->
  ?assertEqual(<<"\"test\"">>, to_val(test)),
  ?assertEqual(<<"\"test\"">>, to_val("test")),
  ?assertEqual(<<"\"test\"">>, to_val(<<"test">>)),
  ?assertEqual(<<"true">>, to_val(true)),
  ?assertEqual(<<"false">>, to_val(false)),
  ?assertEqual(<<"\"te\\ st\"">>, to_val(<<"te st">>)),
  ?assertEqual(<<"\"te\\,st\"">>, to_val(<<"te,st">>)),
  ?assertEqual(<<"\"te\\\"st\"">>, to_val(<<"te\"st">>)),
  ?assertEqual(<<"\"t\\ e\\\"s\\,t\\\"\"">>, to_val(<<"t e\"s,t\"">>)),
  ?assertEqual(<<"123">>, to_val(123)).

to_tag_val_test() ->
  ?assertEqual(<<"test">>, to_tag_val(test)),
  ?assertEqual(<<"test">>, to_tag_val("test")),
  ?assertEqual(<<"test">>, to_tag_val(<<"test">>)),
  ?assertEqual(<<"true">>, to_tag_val(true)),
  ?assertEqual(<<"false">>, to_tag_val(false)),
  ?assertEqual(<<"te\\ st">>, to_tag_val(<<"te st">>)),
  ?assertEqual(<<"te\\,st">>, to_tag_val(<<"te,st">>)),
  ?assertEqual(<<"te\\\"st">>, to_tag_val(<<"te\"st">>)),
  ?assertEqual(<<"t\\ e\\\"s\\,t\\\"">>, to_tag_val(<<"t e\"s,t\"">>)),
  ?assertEqual(<<"123">>, to_tag_val(123)).


encode_time_test() ->
  ?assertEqual(<<" 123">>, encode_time(123)),
  ?assertEqual(<<"">>, encode_time(undefined)).

encode_2_proplist_point_test() ->
  ?assertEqual(
    <<"test 1=1">>,
    encode(test, [{1, 1}])
  ).

encode_2_many_proplist_points_test() ->
  ?assertEqual(
    <<"test 1=1 1\ntest 1=3 2\n">>,
    encode("test", [ [{<<"1">>, 1}], [{<<"1">>, 3}] ], [], 1)
  ).

encode_2_map_point_test() ->
  ?assertEqual(
    <<"test 1=1">>,
    encode(<<"test">>, #{1 => 1})
  ).

encode_2_many_map_points_test() ->
  ?assertEqual(
    <<"test 1=1 1\ntest 1=3 2\n">>,
    encode(<<"test">>, [#{'1' => 1}, #{'1' => 3}], [], 1)
  ).

encode_3_invalid_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(<<"test">>, 'bla-bla', [])
  ).

encode_map_point_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 val=10,text=\"hello\" 123">>,
    encode(
      #{
        measurement => <<"test">>,
        fields => [{"val", 10}, {text, <<"hello">>}],
        tags => [{host, "eu-west"}, {ip, '1.1.1.1'}],
        time => 123
      }
    )
  ).

encode_proplist_point_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 val=10,text=\"hello\" 123">>,
    encode(
      [
        {measurement, <<"test">>},
        {fields, [{"val", 10}, {text, <<"hello">>}]},
        {tags, #{host => "eu-west", ip => '1.1.1.1'}},
        {time, 123}
      ]
    )
  ).

encode_map_points_test() ->
  ?assertEqual(
    <<"test val=10 1\ntest2 val=20 2\n">>,
    encode(
      [
        #{
          measurement => <<"test">>,
          fields => #{"val" => 10},
          time => "1"
        },
        #{
          measurement => <<"test2">>,
          fields => #{"val" => 20},
          time => <<"2">>
        }
      ]
    )
  ).

encode_proplist_points_test() ->
  ?assertEqual(
    <<"test val=10 1\ntest2 val=20 2\n">>,
    encode(
      [
        [
          {measurement, <<"test">>},
          {fields, #{"val" => 10}},
          {time, 1}
        ],
        [
          {measurement, <<"test2">>},
          {fields, #{"val" => 20}},
          {time, 2}
        ]
      ]
    )
  ).

encode_map_without_measurement_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(
      #{
        fields => #{ val => 10}
      }
    )
  ).

encode_map_without_fields_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(
      #{
        measurement => test,
        tags => #{ val => 10}
      }
    )
  ).

encode_proplist_without_measurement_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(
      [
        {fields,  [{val, 10}] }
      ]
    )
  ).

encode_proplist_without_fields_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(
      [
        {measurement, test},
        {tags, #{ val => 10}}
      ]
    )
  ).

encode_at_least_one_invalid_test() ->
  ?assertEqual(
    {error, invalid_data},
    encode(
      [
        [
          {measurement, test},
          {fields, #{ val => 10}}
        ],
        [
          {tags, #{ val => 10}}
        ]
      ]
    )
  ).

encode_3_with_proplist_tags_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 1=1">>,
    encode(<<"test">>, #{1 => 1}, [{host, "eu-west"}, {ip, '1.1.1.1'}])
  ).

encode_3_with_map_tags_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 1=1">>,
    encode(<<"test">>, #{1 => 1}, #{host => 'eu-west', ip => <<"1.1.1.1">>})
  ).

encode_3_with_binary_tags_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 1=1">>,
    encode(<<"test">>, #{1 => 1}, <<",host=eu-west,ip=1.1.1.1">>)
  ).

encode_3_many_points_with_tags_test() ->
  ?assertEqual(
    <<"test,host=eu-west,num=111 memory=\"high\",cpu=20 100\ntest,host=eu-west,num=111 memory=\"low\",cpu=30 101\n">>,
    encode(
      <<"test">>,
      [
        [{memory, "high"}, {"cpu", 20}],
        [{<<"memory">>, "low"}, {cpu, 30}]
      ],
      #{host => 'eu-west', num => 111},
      100
    )
  ).

encode_4_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 1=1 123123">>,
    encode(<<"test">>, #{1 => 1}, [{host, "eu-west"}, {ip, '1.1.1.1'}], 123123)
  ).

encode_4_many_points_test() ->
  ?assertEqual(
    <<"test,host=eu-west,ip=1.1.1.1 1=1,text=\"hello\" 123123">>,
    encode(<<"test">>,
      [[{"1", 1}, {text, <<"hello">>}]],
      [{host, "eu-west"}, {ip, '1.1.1.1'}],
      123123
    )
  ).

encode_escape_tags_test() ->
  ?assertEqual(
    <<"test,host=eu\\ we\\=st,ip=1\\,1\\,1\\,1 1=1">>,
    encode(<<"test">>,
      [[{"1", 1}]],
      [{host, "eu we=st"}, {ip, '1,1,1,1'}]
    )
  ).

encode_escape_fields_test() ->
  ?assertEqual(
    <<"test,host=eu-west 1=1,text=\"hello\\ \\\"man\\=\\=human\\,\\ yo!\\\"\"">>,
    encode(<<"test">>,
      [[{"1", 1}, {text, <<"hello \"man==human, yo!\"">>}]],
      [{host, 'eu-west'}]
    )
  ).

-endif.